duration_extraction/duration_extraction_pipe.py

```diff
@@ -0,0 +1,179 @@
+        return x[0]
+
+
+class DurationExtractionDataset(Dataset):
+
+    def __init__(self,
+                 duration_extractor: DurationExtractor,
+                 paths: Paths,
+                 dataset_ids: List[str],
+                 text_dict: Dict[str, str],
+                 tokenizer: Tokenizer):
+        self.metadata = dataset_ids
+        self.text_dict = text_dict
+        self.tokenizer = tokenizer
+        self.text_dict = text_dict
+        self.duration_extractor = duration_extractor
+        self.paths = paths
+
+    def __getitem__(self, index: int) -> DurationResult:
+        item_id = self.metadata[index]
+        x = self.text_dict[item_id]
+        x = self.tokenizer(x)
+        mel = np.load(self.paths.mel / f'{item_id}.npy')
+        mel = torch.from_numpy(mel)
+        x = torch.tensor(x)
+        att_npy = np.load(str(self.paths.att_pred / f'{item_id}.npy'))
+        att = torch.from_numpy(att_npy)
+        mel_len = mel.shape[-1]
+        mel_len = torch.tensor(mel_len).unsqueeze(0)
+        align_score, _ = attention_score(att.unsqueeze(0), mel_len, r=1)
+        align_score = float(align_score)
+        durs, att_score = self.duration_extractor(x=x, mel=mel, att=att)
+        att_score = float(att_score)
+        durs_npy = durs.cpu().numpy()
+        if np.sum(durs_npy) != mel_len:
+            print(f'WARNINNG: Sum of durations did not match mel length for item {item_id}!')
+        return DurationResult(item_id=item_id, att_score=att_score,
+                              align_score=align_score, durs=durs_npy)
+
+    def __len__(self):
+        return len(self.metadata)
+
+
+class DurationExtractionPipeline:
+
+    def __init__(self,
+                 paths: Paths,
+                 config: Dict[str, Any],
+                 duration_extractor: DurationExtractor) -> None:
+        self.paths = paths
+        self.config = config
+        self.duration_extractor = duration_extractor
+        self.logger = logging.Logger(__name__, level=INFO)
+
+    def extract_attentions(self,
+                           model: Tacotron,
+                           max_batch_size: int = 1) -> float:
+        """
+        Performs tacotron inference and stores the attention matrices as npy arrays in paths.data.att_pred.
+        Returns average attention score.
+
+        Args:
+            model: Tacotron model to use for attention extraction.
+            batch_size: Batch size to use for tacotron inference.
+
+        Returns: Mean attention score. The attention matrices are saved as numpy arrays in paths.att_pred.
+
+        """
+
+        device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
+        model.to(device)
+
+        dataloader = get_binned_taco_dataloader(data_path=self.paths.data, max_batch_size=max_batch_size)
+
+        sum_items = 0
+        sum_att_score = 0
+        pbar = tqdm(dataloader, total=len(dataloader), smoothing=0.01)
```

Comment:
nice using smoothing parameter, I somehow also missed till now that it exists :)